File: cplusplus/Core/VectHor.cpp
/* VectHor.cpp */
#include "VectHor.h"
#include <stdexcept>


//Metodi interni:

//Costruttore di default
VectHor::VectHor(unsigned int n): Vect(1, n) {}

//Costruttore a 2 param
// crea un vettore orizzontale lungo n
// imposta tutti gli elem a init
VectHor::VectHor(unsigned int n, double init): Vect(1, n, init) {}

VectHor* VectHor::clone() const {
	return new VectHor(*this);
}

VectHor* VectHor::sum_p(const MatrixBase* m) const {
	const VectHor* q = dynamic_cast<const VectHor*>(m);
	
	if(!q) //sono permesse somme solo tra un tipo (T1) e un suo sottotipo (T2) (T2<=T1)
		throw std::invalid_argument("è possibile sommare ad un vettore orizzontale solo un vettore orizzontale");
	
	return new VectHor(*this + *q);
}

VectHor* VectHor::dif_p(const MatrixBase* m) const {
	const VectHor* q = dynamic_cast<const VectHor*>(m);
	
	if(!q) //sono permesse differenze solo tra un tipo (T1) e un suo sottotipo (T2) (T2<=T1)
		throw std::invalid_argument("è possibile sottrarre ad un vettore orizzontale solo un vettore orizzontale");
	
	return new VectHor(*this - *q);
}

VectHor* VectHor::sProd_p(double scalar) const {
	return new VectHor((*this) * scalar);
}

VectHor* VectHor::rev_p() const {
	return new VectHor(rev()); //cioè: (*this).rev(); o this->rev();
}

Vect* VectHor::trans_p() const {
	//un VectHor trasposto è un VectVert
	//non posso tornare VectVert* perchè violerei le regole della covarianza
	return new VectVert(trans()); //cioè: (*this).trans(); o this->trans();
}

double& VectHor::at(unsigned int i) {
	return Vect::at(0,i);
}

const double& VectHor::at(unsigned int i) const {
	return Vect::at(0,i);
}

VectHor VectHor::operator-() const {
	return ((*this)*(-1));
}

VectHor VectHor::operator*(double scalar) const {
	VectHor aux = VectHor(numElem());
	for(unsigned int i=0; i<numElem(); ++i)
			aux.at(i) = at(i) * scalar;
	
	return aux;
}

VectVert VectHor::trans() const {
	VectVert aux = VectVert(numElem());
	for(unsigned int i=0; i<numElem(); ++i)
			aux.at(i) = at(i);
	
	return aux;
}

VectHor VectHor::rev() const {
	VectHor aux = VectHor(numElem());
	for(unsigned int i=0; i<numElem(); ++i)
		aux.at(i) = at(numElem()-i-1);
	
	return aux;
}


//Funzioni esterne:

VectHor operator+(const VectHor& v1, const VectHor& v2) {
	if(v1.numElem() != v2.numElem())
		throw std::invalid_argument("la somma è permessa solo tra vettori della stessa lunghezza"); //dimensioni non compatibili
	
	unsigned int e=v1.numElem(); //o: e=v2.numElem();
	VectHor aux = VectHor(e);
	for(unsigned int i=0; i<e; ++i)
		aux.at(i) = v1.at(i) + v2.at(i);
	
	return aux;
}

VectHor operator-(const VectHor& v1, const VectHor& v2) {
	return (v1 + (-v2));
}

VectHor operator*(double scalar, const VectHor& v) {
	return v*scalar;
}

File: java/matrixkalq/VectHor.java
/* VectHor.java */

package matrixkalq;

//import java.lang.IllegalArgumentException;

//Nota: A differenza del C++, in Java non è necessaria alcuna dichiarazione incompleta di classe per poterla utilizzare!


public class VectHor extends Vect {
	
	//Costruttori:
		public VectHor(int n, double init) { super(1,n,init); }
		public VectHor(int n) { super(1, n); }
		public VectHor() { this(0); }
		
	//Override:
		public VectHor clone() {
			return (VectHor)super.clone();
		}
		
	//Metodi 'concretizzati':
		public VectHor sum(MatrixBase m) {
			if(m instanceof VectHor) {
				
				if(numElem() != ((VectHor)m).numElem())
					throw new IllegalArgumentException("la somma è permessa solo tra vettori della stessa lunghezza"); //dimensioni non compatibili
				
				int e=numElem(); //o: e=((VectHor)m).numElem();
				VectHor aux = clone();
				
				for(int i=0; i<e; ++i)
					aux.setElem(i, (getElem(i) + ((VectHor)m).getElem(i)));
				
				return aux;
			}
			else { //sono permesse somme solo tra un tipo (T1) e un suo sottotipo (T2) (T2<=T1)
				throw new IllegalArgumentException("è possibile sommare ad un vettore orizzontale solo un vettore orizzontale");
			}
		}
		
		public VectHor dif(MatrixBase m) {
			if(m instanceof VectHor) {
				return (this.sum(m.sProd(-1)));
			}
			else { //sono permesse differenze solo tra un tipo (T1) e un suo sottotipo (T2) (T2<=T1)
				throw new IllegalArgumentException("è possibile sottrarre ad un vettore orizzontale solo un vettore orizzontale");
			}
		}
		
		public VectHor sProd(double scalar) {
			VectHor aux = clone();
			
			for(int i=0; i<numElem(); ++i)
				aux.setElem(i, (getElem(i) * scalar));
			
			return aux;
		}
		
		public VectHor rev() {
			VectHor aux = clone();
			
			for(int i=0; i<numElem(); ++i)
				aux.setElem(i, getElem(numElem()-i-1));
			
			return aux;
		}
		
		//Un VectHor trasposto è un VectVert
		public VectVert trans() {
			VectVert aux = new VectVert(numElem());
			
			for(int i=0; i<numElem(); ++i)
				aux.setElem(i, getElem(i));
			
			return aux;
		}
		
		public double getElem(int i) {
			return getElem(0,i); //o: return super.getElem(0,i);
		}
		
		public void setElem(int i, double val) {
			setElem(0,i,val); //o: super.setElem(0,i,val);
		}
		
	//Metodi ereditati:
		//	public int numRows();
		//	public int numCols();
		//	public boolean equals(Object obj);
		
		//	public String toString();
		//	public double getElem(int i, int j);
		//	public void setElem(int i, int j, double val);
		//	public void swapRows(int row1, int row2, int fromCol);
		//	public void swapCols(int col1, int col2, int fromRow);
		//	public void print(DataOutputStream os, String separator) throws IOException;
		//	public void load(DataInputStream is) throws EOFException, IOException;
		//	public void fill(double val);
		//	public void setRandInt(int from, int to);
		//	public void setRandReal(double from, double to);
		
		//	public int rank();
		//	public int numElem();
		//	public void swapElem(int e1, int e2);
		
	//Eccezioni:
		//	ArrayIndexOutOfBoundsException: getElem(..), setElem(..), swapRows(..), swapCols(..)
		//	NumberFormatException: load(..)
		//	IllegalArgumentException: load(..), setRandInt(..), setRandReal(..)
		//	RuntimeException: clone()
		
		//	ArrayIndexOutOfBoundsException: swapElem(..) [a causa di getElem(int i) e/o setElem(int i, double val)]
		
		//	IllegalArgumentException: sum(..), dif(..)
		//	ArrayIndexOutOfBoundsException: getElem(..), setElem(..)
		
} //fine VectHor
